lib/timers.js

```diff
@@ -62,6 +62,37 @@ function listOnTimeout() {
   var diff, first, threw;
   while (first = L.peek(list)) {
     diff = now - first._idleStart;
+    // This handles the case of a timer that was created within a timers
+    // callback with the same timeout value. For instance, when processing the
+    // timer that would call `bar` in such code:
+    //
+    //   setTimeout(function foo() { setTimeout(function bar() {}, 0) }, 0);
+    //
+    // or
+    //
+    //   setTimeout(function foo() { setTimeout(function bar() {}, 500) }, 500);
+    //
+    // We want to make sure that newly added timer fires in the next turn of the
+    // event loop at the earliest. So even if it's already expired now,
+    // reschedule it to fire later.
+    //
+    // At that point, it's not necessary to process any other timer in that
+    // list, because any remaining timer has been added within a callback of a
+    // timer that has already been processed, and thus needs to be processed at
+    // the earliest not in the current tick, but when the rescheduled timer will
+    // expire.
+    //
+    // See: https://github.com/joyent/node/issues/25607
+    if (now <= first._idleStart) {
```

Comment:
`first._idleStart` is the time at which a timer was scheduled. Since `now` is updated to be the time at the start of `listOnTimeout` execution, `now` is always `> first._idleStart`, _unless_ `first` was scheduled from within the call to `listOnTimeout`.